clar2wasm/src/wasm_utils.rs

```diff
@@ -1285,17 +1286,22 @@ pub fn call_function<'a>(
     // Call the function
     func.call(&mut store, &wasm_args, &mut results)
         .map_err(|e| {
-            // TODO: If the root cause is a clarity error, we should be able to return that,
-            //       but it is not cloneable, so we can't return it directly.
-            //       If the root cause is a trap from our Wasm code, then we need to translate
-            //       it into a Clarity error.
-            //       See issue stacks-network/clarity-wasm#104
-            // if let Some(vm_error) = e.root_cause().downcast_ref::<crate::vm::errors::Error>() {
-            //     vm_error.clone()
-            // } else {
-            //     Error::Wasm(WasmError::Runtime(e))
-            // }
-            Error::Wasm(WasmError::Runtime(e))
+            if let Some(vm_error) = e.root_cause().downcast_ref::<Error>() {
+                // This unsafe operation returns the value of a location pointed by `*mut T`.
+                // The purpose of this code is to take the ownership of the `vm_error` value.
+                // The replaced `T` value is deallocated after the operation. Therefore, the chosen `T`
+                // is a dummy value, solely to satisfy the signature of the replace function
+                // and not cause harm when it is deallocated.
+                // Specifically, Error::Wasm(WasmError::ModuleNotFound) was selected as the placeholder value.
+                return unsafe {
+                    core::ptr::replace(
+                        (vm_error as *const Error) as *mut Error,
+                        Error::Wasm(WasmError::ModuleNotFound),
+                    )
+                };
+            }
+
+            error_mapping::from_runtime_error_code(instance, &mut store)
```

Comment:
I'm really not sure we should call this function here. I would say that `from_runtime_error_code` is used when we want to get the error from a faulty code to a clarity function, but here it could return anything, like a _wasmtime_ error.
My opinion is that this function should remain `Error::Wasm(WasmError::Runtime(e))` here, because an error returned by error code would be caught in the `if let`.